Implement new packages auth, azure and git for passwordless authentication scenarios

dipti-pai opened the pull request:
Changes include partial implementation of [RFC-007](https://github.com/fluxcd/flux2/tree/main/rfcs/0007-git-repo-passwordless-auth) for Azure provider.

- `azure` sub-package in `auth` with API to get azure devops access token using workload identity and associated token provider unit tests.
- Changes to `git` and `gogit` package using the API to get credentials invoking provider specific implementation (only Azure is added currently) and associated git credentials/gogit client unit tests.

dipti-pai commented on auth/azure/token_provider.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+const (
+	AzureDevOpsRestApiScope = "499b84ac-1321-427f-aa17-267ca6975798/.default"
```
This Guid is the application id of the Azure DevOps app registration from Microsoft's tenant. Every Entra ID tenant has an enterprise application linked to this app registration, thus this guid is common. To verify this, we can run the below command "az ad sp show --id 499b84ac-1321-427f-aa17-267ca6975798". In the output, it lists alternate service principal names for the application. 

```
  "servicePrincipalNames": [
    "499b84ac-1321-427f-aa17-267ca6975798",
    "https://wcus0.app.vssps.visualstudio.com",
    **"https://app.vssps.visualstudio.com/",**
    "https://app.vssps.vsallin.net/",
    "https://analytics.dev.azure.com/"
  ],
```
Replacing this guid with the name from the output above (https://app.vssps.visualstudio.com/) works as well to get the correct access token. However, Microsoft documentation uses the GUID as described [here](https://learn.microsoft.com/en-us/rest/api/azure/devops/tokens/?view=azure-devops-rest-7.1&tabs=powershell) and [here](https://learn.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/service-principal-managed-identity?view=azure-devops#q-can-i-use-a-service-principal-or-managed-identity-with-azure-cli). The example to do git clone using this bearer token is [here](https://learn.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/service-principal-managed-identity?view=azure-devops#q-can-i-use-a-service-principal-to-do-git-operations-like-clone-a-repo). Will update the code with these documentation links and a descriptive comment. 

Getting the ARM token as in the previous attempt in the common auth package did not return a valid token that worked with Azure DevOps.

dipti-pai commented on auth/azure/token_provider.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+func (p *Provider) GetToken(ctx context.Context) (*azcore.AccessToken, error) {
+	if len(p.scopes) == 0 {
+		return nil, fmt.Errorf("error scopes must be specified")
+	}
```
I had thought about this while implementing the code. Using the azure public cloud scope does not work for Azure DevOps, but if we intend to share this package with ACR login for oci, it may be ok to default to azure public cloud and make sure that all git code with provider Azure overrides this to use the AzureDevOpsScope. I thought we can make the change for the default scope if/when we start using the package with oci. 

Please let me know your take on this and if you agree it is ok to default to azure public cloud scope in a subsequent PR when oci uses this package.

dipti-pai commented on auth/git/credentials.go:
```diff
@@ -0,0 +1,109 @@
+	var creds Credentials
+	log := log.FromContext(ctx)
+
+	if authOpts.Cache != nil {
+		creds, exists, err := getObjectFromCache(authOpts.Cache, url)
+		if err != nil {
+			log.Error(err, "failed to get credential object from cache")
+		}
+		if exists {
+			return &creds, nil
+		}
+	}
```
Small clarification here: I assume "git client" in the above comment means source-controller since that's the one that will call GetCredentials() and not the gogit client in pkg? Could you clarify? Thanks!

dipti-pai commented on auth/azure/token_provider.go:
```diff
@@ -0,0 +1,84 @@
+specific language governing permissions and limitations under the License.
+*/
+
+package azure
```
The common code with the ACR package will be the token_provider with different options for scopes (public cloud, china, us gov cloud, etc). These scope options are not added yet. ACR package would call token provider followed by ACR specific code for token exchange for ACR.

souleb commented on auth/git/credentials.go:
```diff
@@ -0,0 +1,110 @@
+	return data
+}
+
+func CacheCredentials(ctx context.Context, url string, authOpts *AuthOptions, creds *Credentials) {
```
Is this necessary? The caller provides the cache. If we want to verify the `token` before caching it, we can do that directly from the caller without relying on a function in this pkg. This is true as well if we go forward and decide that the client is solely responsible for caching.

dipti-pai commented:
I removed all the caching-related code from the PR as I wanted to get thoughts about moving some files around before implementing. 

With GetCredentials() call moved to gogit client in the new function CloneWithProvider and get/set/delete from cache would also be called from gogit client, would it make sense to not have a separate git package in pkg/auth/git rather move credentials.go to to existing git package at pkg/git and add the new cache.go to pkg/git with methods to get/set/delete entries from cache. 

- AuthOptions (in credentials.go right now) will be renamed to ProviderAuthOptions and expanded to hold the cache passed by the caller. This can be moved to pkg/git/options.go which has other AuthOptions defined. 
- CloneWithProvider in client.go will be expanded with the following logic (pseudo code below)

function CloneWithProvider() {
    if getCredentialsFromCache != nil {
           commit, err = Clone()
           if err == UnAuthorized {
              invalidateCache() 
           } else {
               return commit, err
           }
    }
    
    call getCredentials API to fetch new token
    Clone()
}

stefanprodan commented on git/gogit/client.go:
```diff
@@ -38,9 +40,7 @@ import (
-
-	"github.com/fluxcd/pkg/git"
-	"github.com/fluxcd/pkg/git/repository"
+	"sigs.k8s.io/controller-runtime/pkg/log"
```
We can't have controller-runtime as a dependency here, this is a major blocker for me.

stefanprodan commented on auth/go.mod:
```diff
@@ -0,0 +1,27 @@
+	github.com/Azure/azure-sdk-for-go/sdk/azcore v1.12.0
+	github.com/Azure/azure-sdk-for-go/sdk/azidentity v1.7.0
+	github.com/onsi/gomega v1.33.1
+	k8s.io/utils v0.0.0-20240502163921-fe8a2dddb1d0
```
Please get rid of this dependency, the Git package has to say free of Kubernetes deps.

stefanprodan commented on git/gogit/client.go:
```diff
@@ -251,6 +251,49 @@ func (g *Client) Clone(ctx context.Context, url string, cfg repository.CloneConf
+		// If cache is specified, try cloning with cached credential if it exists,
+		// else fallback to fetching credentials from provider.
+		if g.authOpts.Cache != nil {
+			cachedCreds, exists, err := getObjectFromCache(g.authOpts.Cache, url)
```
What if there are hundreds of DevOps repos in the same Azure account, shouldn't we cache the token based on the Azure account instead of using the repo URL?

stefanprodan replied:
Ok we can add the UID to the client options and use it as the cache key, while this will avoid conflicts it doesn't make Azure DevOps OIDC fit for multi-tenancy as there is no way to restrict a tenant from accessing another tenant repos.

pjbgf replied:
> In this per object identity case, I think we can't simplify anything and have to use the full repo URL as cache key.

Heavy plus one on this, as different identities may have different permissions to the same repository.
An additional point of caution here is cache invalidation when the object identity is removed/changed, which makes me think that potentially concatenating both identity metadata plus the UID may be a sensible approach.

dipti-pai replied:
Thanks for the discussions. 

The reason I had kept the URL as the cache key was because of the way fine-grained permissions can be assigned at a per-project/per-repository level in Azure DevOps. From source controller's perspective, since there is a single (workload) identity that can be setup, in a happy path, this identity would need to be given access to all repositories that have the `provider` field in `GitRepository` spec set to `azure` to work correctly. The token associated with this identity would be fetched and cached once per repository. If the user sets things up incorrectly, for example: where there are multiple repositories and the identity is not given access to one of the repositories that has `provider` set to `azure`, then only the incorrectly configured repository fails with authentication error (as it should). The cached token is still valid for the repositories where permissions are setup correctly. 

Using `GitRepository` object's UID instead of repo URL should behave similarly and can be used as a cache key as well if this makes the approach common for other providers (Github, Gitlab). I can make the changes needed for this. 

Regarding cache invalidation, the approach is to treat the cache as best-effort. If there are any errors encountered while using the cached credentials, we fall back to fetching and using the credentials from the provider. If there are errors encountered while using the cached credential or while interacting with the cache (set/get), these are logged for debugging purposes but not propagated as user-facing errors.

stefanprodan replied:
> different identities may have different permissions to the same repository

@pjbgf how can you have different identities with Workload ID?

stefanprodan commented on git/gogit/client.go:
```diff
@@ -251,6 +251,49 @@ func (g *Client) Clone(ctx context.Context, url string, cfg repository.CloneConf
+	commit, err := g.clone(ctx, url, cfg)
+	if err == nil && providerCreds != nil && g.authOpts.Cache != nil {
+		// Cache credentials if clone was successful with provider credentials
+		cacheObject(g.authOpts.Cache, *providerCreds, url, expiresOn)
```
There is an unhandled error here, the cache can be full.

dipti-pai commented:
Manually validated when proxy server configuration is specified in GitRepository `.spec.proxySecretRef`, the requests to acquire azure devops access token go through the proxy

Relevant logs from source-controller and proxy server are below 

```
{"level":"info","ts":"2024-09-06T07:46:29.767Z","msg":"no changes since last reconcilation: observed revision 'main@sha1:c629e6ee0520bb2c9328addb5229621fa0f31eab'","controller":"gitrepository","controllerGroup":"source.toolkit.fluxcd.io","controllerKind":"GitRepository","GitRepository":{"name":"azure-devops","namespace":"default"},"namespace":"default","name":"azure-devops","reconcileID":"bc7870d0-fab9-4c94-aafc-fcd2419a4b76"}
{"level":"info","ts":"2024-09-06T07:47:28.875Z","msg":"no changes since last reconcilation: observed revision 'main@sha1:c629e6ee0520bb2c9328addb5229621fa0f31eab'","controller":"gitrepository","controllerGroup":"source.toolkit.fluxcd.io","controllerKind":"GitRepository","GitRepository":{"name":"azure-devops","namespace":"default"},"namespace":"default","name":"azure-devops","reconcileID":"e88474a2-095e-4e69-beb2-9cefb0bce263"}
{"level":"info","ts":"2024-09-06T07:48:29.490Z","msg":"no changes since last reconcilation: observed revision 'main@sha1:c629e6ee0520bb2c9328addb5229621fa0f31eab'","controller":"gitrepository","controllerGroup":"source.toolkit.fluxcd.io","controllerKind":"GitRepository","GitRepository":{"name":"azure-devops","namespace":"default"},"namespace":"default","name":"azure-devops","reconcileID":"44368d25-ab9a-47e5-8481-a94150d8be74"}
{"level":"info","ts":"2024-09-06T07:49:31.109Z","msg":"no changes since last reconcilation: observed revision 'main@sha1:c629e6ee0520bb2c9328addb5229621fa0f31eab'","controller":"gitrepository","controllerGroup":"source.toolkit.fluxcd.io","controllerKind":"GitRepository","GitRepository":{"name":"azure-devops","namespace":"default"},"namespace":"default","name":"azure-devops","reconcileID":"7301e23e-0e2d-41ae-88dd-9b4949a450d2"}
```

```
2024/09/06 07:46:28 [217] INFO: Running 0 CONNECT handlers
2024/09/06 07:46:28 [217] INFO: Accepting CONNECT to login.microsoftonline.com:443
2024/09/06 07:46:29 [218] INFO: Running 0 CONNECT handlers
2024/09/06 07:46:29 [218] INFO: Accepting CONNECT to dev.azure.com:443
2024/09/06 07:47:28 [219] INFO: Running 0 CONNECT handlers
2024/09/06 07:47:28 [219] INFO: Accepting CONNECT to login.microsoftonline.com:443
2024/09/06 07:47:28 [220] INFO: Running 0 CONNECT handlers
2024/09/06 07:47:28 [220] INFO: Accepting CONNECT to dev.azure.com:443
2024/09/06 07:48:28 [221] INFO: Running 0 CONNECT handlers
2024/09/06 07:48:28 [221] INFO: Accepting CONNECT to login.microsoftonline.com:443
2024/09/06 07:48:28 [222] INFO: Running 0 CONNECT handlers
2024/09/06 07:48:28 [222] INFO: Accepting CONNECT to dev.azure.com:443
2024/09/06 07:49:30 [223] INFO: Running 0 CONNECT handlers
2024/09/06 07:49:30 [223] INFO: Accepting CONNECT to login.microsoftonline.com:443
2024/09/06 07:49:30 [224] INFO: Running 0 CONNECT handlers
2024/09/06 07:49:30 [224] INFO: Accepting CONNECT to dev.azure.com:443
```

stefanprodan commented on git/go.mod:
```diff
@@ -1,20 +1,35 @@
-go 1.22.0
+go 1.22.4
+
+toolchain go1.22.5
```
Please remove the `toolchain` line from all go.mod files and run `make tidy`.

dipti-pai commented on git/credentials_test.go:
```diff
@@ -0,0 +1,107 @@
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			str := ""
+			ctx := context.WithValue(context.TODO(), "scope", &str)
```
Removed the context scope check from mock credential and from git. The scope validation tests are done in auth/azure/client_test.go

dipti-pai commented on git/credentials.go:
```diff
@@ -0,0 +1,73 @@
+
+// GetCredentials returns authentication credentials for accessing the provided
+// Git repository.
+func GetCredentials(ctx context.Context, url string, providerOpts *ProviderOptions) (*Credentials, time.Time, error) {
```
Thanks for catching. It was used when caching was done in credentials and got missed in the refactoring.

stefanprodan reviewed (approved):
LGTM

Thanks @dipti-pai 🥇
